_posts/2022-03-07-running-clojure-on-constrained-memory.md

```diff
@@ -0,0 +1,529 @@
+metrics collection to fail to work on any operating system not based
+on Linux. The discussion on virtual memory is also more or less based
+on my understanding of Linux.
+
+# Effects of small memory
+
+Limiting the amount of main memory on a system that runs the JVM
+has two important consequences that I wish to point out. First, a direct
+effect that happens regardless of if our program runs on the JVM or some other
+runtime, is that the operating system must provide its services with
+less memory. This affects especially the page cache, but it also may result
+in swapping-out of memory content. Even in absence of swap, heavy file system
+read load can result from having to reclaim cached-in file system content
+too soon. Finally, if the memory reclamation mechanisms fail to
+otherwise find memory needed for the operating system to operate, it
+will be necessary to terminate some memory-intensive process or even
+restart the whole system. Both actions are disruptive, but normally we really
+want to be way below the level of memory use where the memory
+reclamation mechanism becomes inefficient.
+
+The second important effect from a small main memory is the JVM heap
+size. To avoid driving the operating system into a memory-starved
+state, at least OpenJDK defaults to only using half of the main
+memory for its heap. This is a quite sensible default, but it is not
+always the best value. The JVM process needs memory for quite a few
+things other than just the application heap, so the remaining half may
+be insufficient. The operating system might also be running some other
+programs that require memory. On the other hand, if the workload of
+the operating system is lean
+enough, it may be advantageous to use more than half of the main memory
+for the JVM heap.
+
+I suppose that most people using JVM based programming languages are
+familiar with the heap size of JVM being manually tunable. On the
+other hand, at least I previously had only vague ideas on what exactly I
+should pick on a small-memory system. Selecting a small heap will make
+it less likely that things apart from the JVM run out of memory, but
+it at the very least increases the frequency of GC runs. More
+severely, it may also happen that the GC finds nothing to free, not
+unlike how the operating system might find itself unable to reclaim
+memory. In this case, a new memory allocation inside the JVM likely
+throws an
+[OutOfMemoryError](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/OutOfMemoryError.html).
+This error can in theory be caught and handled, but it is difficult to do it
+reliably or even figure out a sensible course of action. Doing almost
+anything is likely to require further allocations from the heap.
+
+To summarize, it seems that the range of reasonable values for the
+heap size is quite large. The low end of the range is
+constrained at least by the risk of having the program terminate on an
```

Comment:
an OOME wouldn't terminate the whole JVM program, just the thread, but might leave your program in an inconsistent state (especially if you are using mutable objects).

Reply:
Thank you for pointing this out. I wasn't fully conscious that it would only end the thread. But there is a big difference between a terminated program and broken but running program, so should fix this.